pkg/upgrade/control_plane.go

```diff
@@ -1251,18 +1373,46 @@ func (u *ControlPlaneUpgrader) waitForMatchingNode(ctx context.Context, rawProvi
 	return &matchingNode, nil
 }
 
-func (u *ControlPlaneUpgrader) waitForNodeReady(ctx context.Context, newNode *v1.Node) error {
+func (u *ControlPlaneUpgrader) waitForNodeReady(ctx context.Context, node *v1.Node) error {
 	// wait for NodeReady
 	err := wait.PollImmediateUntil(15*time.Second, func() (bool, error) {
-		ready := u.isReady(newNode.Name)
+		ready := u.isReady(node.Name)
 		return ready, nil
 	}, ctx.Done())
 	if err != nil {
-		return errors.Wrapf(err, "components on node %s are not ready", newNode.Name)
+		return errors.Wrapf(err, "components on node %s are not ready", node.Name)
 	}
 	return nil
 }
 
+func (u *ControlPlaneUpgrader) waitForNodeWithMasterLabel(ctx context.Context, node *v1.Node) error {
+	err := wait.PollImmediateUntil(15*time.Second, func() (bool, error) {
+		return u.hasMasterNodeLabel(node.Name), nil
+	}, ctx.Done())
+	if err != nil {
+		return errors.Wrapf(err, "node %s does not have master label %q", node.Name, LabelNodeRoleMaster)
+	}
+	return nil
+}
+
+func (u *ControlPlaneUpgrader) hasMasterNodeLabel(nodeName string) bool {
+	log := u.log.WithValues("node", nodeName)
+	log.Info("Verifying node has master label", "label", LabelNodeRoleMaster)
+
+	node, err := u.targetKubernetesClient.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		log.Info("Node not found yet")
+		return false
+	} else if err != nil {
```

Comment:
nit: because you have a `return` inside the preceding `if` block, you don't need an `else` here